include/session/session_protocol.hpp

```diff
@@ -0,0 +1,307 @@
+#pragma once
+
+#include <session/session_protocol.h>
+
+#include <session/config/pro.hpp>
+#include <session/types.hpp>
+#include <span>
+
+/// A complimentary file to session encrypt (which has the low level encryption function for Session
+/// protocol types). This file contains high-level helper functions for decoding payloads on the
+/// Session protocol. Prefer functions here before resorting to the lower-level cryptography.
+
+// NOTE: In the CPP file we use C-style enums for bitfields and CPP-style enums for non-bitfield
+// enums where we can to benefit from the type-safety of strong enums.
+//
+// CPP doesn't support named bitfields without casting or operator overloads but C-style
+// enums support it very well. The only issue is that using a native C-style enum enforces some type
+// restrictions that compilers dislike when attempting to manipulate bit fields. For example:
+//
+//   enum Feature {x = 1 << 0, y = 1 << 1}
+//   Feature f = x | y
+//
+// Causes the compiler to complain about trying to do bit ops/assign an unsigned integer to an enum
+// `Feature`. We use a common C pattern/trick by suffixing an underscore to the the original enum,
+// then type define the non-suffixed enum to an unsigned integer:
+//
+//   enum Feature_ {x = 1 << 0, y = 1 << 1}
+//   typedef U64 Feature
+//   Feature f = x | y
+//
+// Does not trigger errors as the underlying type of `f` is actually an unsigned integer. The type
+// define is merely a hint to the user to what flags are to be used when manipulating the variable.
+
+namespace session {
+
+namespace config::groups {
+    class Keys;
+}
+
+enum class ProStatus {
+    // Proof not set
+    Nil = PRO_STATUS_NIL,
+    // Proof set; pro proof sig was not produced by the Pro backend key
+    InvalidProBackendSig = PRO_STATUS_INVALID_PRO_BACKEND_SIG,
+    // Proof set; envelope pro sig was not produced by the Rotating key
+    InvalidUserSig = PRO_STATUS_INVALID_USER_SIG,
+    // Proof set, is verified; has not expired
+    Valid = PRO_STATUS_VALID,
+    // Proof set, is verified; has expired
+    Expired = PRO_STATUS_EXPIRED,
+};
```

Comment:
Should we leave the expired state out of this status? Whether a pro has expired depends on current time, which I find it the best to leave it to the caller to decide - so that:
1. They can supply their own clock
2. They know when to recheck for validity. 

For this reason I suggest we do these states instead:


```suggestion
enum class ProStatus {
    // Proof not set
    Nil = PRO_STATUS_NIL,
    // Proof set; pro proof sig was not produced by the Pro backend key
    InvalidProBackendSig = PRO_STATUS_INVALID_PRO_BACKEND_SIG,
    // Proof set; envelope pro sig was not produced by the Rotating key
    InvalidUserSig = PRO_STATUS_INVALID_USER_SIG,
    // Proof set, is verified; whether it's expired you should check the expiration time in pro_proof
    Verified = PRO_STATUS_VERIFIED,
};
```

Thoughts?

Reply:
Yeah this is currently the case that the caller is in control of the clock.
```cpp
/// - `unix_ts` -- pass in the current system time in seconds which is used to determine, whether or
///   not the Session Pro proof has expired or not if it is in the payload. Ignored if there's no
///   proof in the message.

DecryptedEnvelope decrypt_envelope(
        const DecryptEnvelopeKey& keys,
        std::span<const uint8_t> envelope_payload,
        std::chrono::sys_seconds unix_ts,
        const array_uc32& pro_backend_pubkey);
```
Pro verification in regards to extracting a pro status from the proof is currently embedded in the decryption function but I intend to pull that out to some utility functions some point later so you can call it standalone. I plan to add those after if we're generally happy with the approach here before writing all the helper utilities that go ontop.

Is that siufficient for your use case? Or would you still want to separate the expired-ness from PRO_STATUS.

Reply:
The use case is when `decrypt_envelop` is called, the user hasn't expired yet, let's say 10s later it expires, we won't be calling the `decrypt_envelop` to determine the validity again, we (the application code) will have to arrange its own timer to check the expire timestamp - hence it's best not for the libsession to say whether the pro has expired as it will be tempting to think that as long as the result you get is valid, it's valid forever. Just pushing the expiration check's responsibility to the user forces them to handle it correctly.

This is more about making the API hard to misuse, as I believe at the current form, all the data the application needs is right there. In terms of extracting a pro status from the proof - I think I have some other ideas on the `DecryptedEnvelop` so let's see how you think of that suggestion instead.

Reply:
After a thought I think it doesn't make sense to not have libsession validating the expired status, "expiration" after all, is a criteria of validity. So forget about this one

Reply:
Sure, yes I think so we can move expiry out and not merge it into the status field.

Reply:
This has been handled out-of-band, we've decided to keep expired in the status, but, reaffirmed that it's useful to have that separated as the clients envision having to check expiry themselves as well independently from decryption at later points.